packages/tools/src/utilities/contourROITool/Interpolation/freehandInterpolate/interpolate.ts

```diff
@@ -0,0 +1,656 @@
+
+  const indicies = c1HasMoreNodes ? c1ir.I : c2ir.I;
+
+  for (let i = 0; i < c1ir.x.length; i++) {
+    if (indicies[i]) {
+      cInterp.x.push(c1ir.x[i] + (c2ir.x[i] - c1ir.x[i]) * zInterp);
+      cInterp.y.push(c1ir.y[i] + (c2ir.y[i] - c1ir.y[i]) * zInterp);
+      cInterp.z.push(c1ir.z[i] + (c2ir.z[i] - c1ir.z[i]) * zInterp);
+    }
+  }
+
+  return cInterp;
+}
+
+/**
+ * _generateInterpolationContourPair - generates two aligned contours with an
+ * equal number of nodes from which an intermediate contour may be interpolated.
+ *
+ * @param c1 - The first contour.
+ * @param c2 - The second contour.
+ * @returns -  An object containing the two contours.
+ */
+function _generateInterpolationContourPair(c1, c2) {
+  const cumPerim1 = _getCumulativePerimeter(c1);
+  const cumPerim2 = _getCumulativePerimeter(c2);
+
+  const interpNodes = Math.max(
+    Math.ceil(cumPerim1[cumPerim1.length - 1] / dP),
+    Math.ceil(cumPerim2[cumPerim2.length - 1] / dP)
+  );
+
+  const cumPerim1Norm = _normalisedCumulativePerimeter(cumPerim1);
+  const cumPerim2Norm = _normalisedCumulativePerimeter(cumPerim2);
+
+  const numNodes1 = interpNodes + c2.x.length;
+  const numNodes2 = interpNodes + c1.x.length;
+
+  // concatinate p && cumPerimNorm
+  const perim1Interp = _getInterpolatedPerim(numNodes1, cumPerim1Norm);
+  const perim2Interp = _getInterpolatedPerim(numNodes2, cumPerim2Norm);
+
+  const perim1Ind = _getIndicatorArray(c1, numNodes1);
+  const perim2Ind = _getIndicatorArray(c2, numNodes2);
+
+  const nodesPerSegment1 = _getNodesPerSegment(perim1Interp, perim1Ind);
+  const nodesPerSegment2 = _getNodesPerSegment(perim2Interp, perim2Ind);
+
+  const c1i = _getSuperSampledContour(c1, nodesPerSegment1);
+  const c2i = _getSuperSampledContour(c2, nodesPerSegment2);
+
+  // Keep c2i fixed and shift the starting node of c1i to minimise the total length of segments.
+  _shiftSuperSampledContourInPlace(c1i, c2i);
+
+  return _reduceContoursToOriginNodes(c1i, c2i);
+}
+
+/**
+ * _reduceContoursToOriginNodes - Removes any nodes from the contours that don't
+ * correspond to an original contour node.
+ *
+ * @param c1i - The first super-sampled contour.
+ * @param c2i - The second super-sampled contour.
+ * @returns  An object containing the two reduced contours.
+ */
+function _reduceContoursToOriginNodes(c1i, c2i) {
+  const c1Interp = {
+    x: [],
+    y: [],
+    z: [],
+    I: [],
+  };
+  const c2Interp = {
+    x: [],
+    y: [],
+    z: [],
+    I: [],
+  };
+
+  for (let i = 0; i < c1i.x.length; i++) {
+    if (c1i.I[i] || c2i.I[i]) {
+      c1Interp.x.push(c1i.x[i]);
+      c1Interp.y.push(c1i.y[i]);
+      c1Interp.z.push(c1i.z[i]);
+      c1Interp.I.push(c1i.I[i]);
+
+      c2Interp.x.push(c2i.x[i]);
+      c2Interp.y.push(c2i.y[i]);
+      c2Interp.z.push(c2i.z[i]);
+      c2Interp.I.push(c2i.I[i]);
+    }
+  }
+
+  return {
+    c1Interp,
+    c2Interp,
+  };
+}
+
+/**
+ * _shiftSuperSampledContourInPlace - Shifts the indicies of c1i around to
```

Comment:
Could you move the helper functions like this one into a utilities.interpolation package?  They are extremely helpful, and having them be separate files allows for other uses of them.
I really like what you are doing here, just that I want to use it in more places.  Consider being able to interpolate between an ellipse and a freeform, or a contour generated on threshold values, with the threshold guiding the interpolation.  lots of very useful use cases here.